From ddfa58eb53c0a6ce3186827e106fe3bdbb682735 Mon Sep 17 00:00:00 2001
From: James Eidson <jmes@google.com>
Date: Wed, 18 Feb 2026 14:15:10 -0800
Subject: [PATCH 5/9] Revert "[nfc] Fix string injection in default payment app
 selector"

This reverts commit b04265c1b84104884654d4957c5fb3a8ac84bc00.

Reason for revert: revert needed to fix regression

Bug: 429417453
Change-Id: Ieead4198a61009a048b8336377febff3c123cd57
---
 .../settings/nfc/DefaultPaymentSettings.java  | 29 +++++++++----------
 .../android/settings/nfc/PaymentBackend.java  | 19 ------------
 2 files changed, 14 insertions(+), 34 deletions(-)

diff --git a/src/com/android/settings/nfc/DefaultPaymentSettings.java b/src/com/android/settings/nfc/DefaultPaymentSettings.java
index b5057ed64f2..d75844bb0ff 100644
--- a/src/com/android/settings/nfc/DefaultPaymentSettings.java
+++ b/src/com/android/settings/nfc/DefaultPaymentSettings.java
@@ -17,6 +17,7 @@
 package com.android.settings.nfc;
 
 import android.app.settings.SettingsEnums;
+import android.content.ComponentName;
 import android.content.Context;
 import android.content.Intent;
 import android.graphics.drawable.Drawable;
@@ -44,9 +45,6 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
-import java.util.Map;
-import java.util.function.Function;
-import java.util.stream.Collectors;
 
 /**
  * DefaultPaymentSettings handles the NFC default payment app selection.
@@ -55,7 +53,7 @@ public class DefaultPaymentSettings extends DefaultAppPickerFragment {
     public static final String TAG = "DefaultPaymentSettings";
 
     private PaymentBackend mPaymentBackend;
-    private Map<String, PaymentAppInfo> mAppInfos;
+    private List<PaymentAppInfo> mAppInfos;
     private FooterPreference mFooterPreference;
 
     @Override
@@ -69,19 +67,22 @@ public class DefaultPaymentSettings extends DefaultAppPickerFragment {
     }
 
     @Override
-    @SuppressWarnings("NullAway")
     protected String getDefaultKey() {
         PaymentAppInfo defaultAppInfo = mPaymentBackend.getDefaultApp();
-        if (defaultAppInfo == null) return null;
-        return defaultAppInfo.getKey();
+        if (defaultAppInfo != null) {
+            return defaultAppInfo.componentName.flattenToString() + " "
+                    + defaultAppInfo.userHandle.getIdentifier();
+        }
+        return null;
     }
 
     @Override
     protected boolean setDefaultKey(String key) {
-        PaymentAppInfo appInfo = mAppInfos.get(key);
-        if (appInfo == null) return true;
-        mPaymentBackend.setDefaultPaymentApp(
-                appInfo.componentName, appInfo.userHandle.getIdentifier());
+        String[] keys = key.split(" ");
+        if (keys.length >= 2) {
+            mPaymentBackend.setDefaultPaymentApp(ComponentName.unflattenFromString(keys[0]),
+                    Integer.parseInt(keys[1]));
+        }
         return true;
     }
 
@@ -89,9 +90,7 @@ public class DefaultPaymentSettings extends DefaultAppPickerFragment {
     public void onAttach(Context context) {
         super.onAttach(context);
         mPaymentBackend = new PaymentBackend(getActivity());
-        mAppInfos = mPaymentBackend.getPaymentAppInfos()
-                .stream()
-                .collect(Collectors.toMap(PaymentAppInfo::getKey, Function.identity()));
+        mAppInfos = mPaymentBackend.getPaymentAppInfos();
     }
 
     @Override
@@ -148,7 +147,7 @@ public class DefaultPaymentSettings extends DefaultAppPickerFragment {
     @Override
     protected List<? extends CandidateInfo> getCandidates() {
         final List<NfcPaymentCandidateInfo> candidates = new ArrayList<>();
-        for (PaymentAppInfo appInfo: mAppInfos.values()) {
+        for (PaymentAppInfo appInfo: mAppInfos) {
             UserManager um = getContext().createContextAsUser(
                     appInfo.userHandle, /*flags=*/0).getSystemService(UserManager.class);
             boolean isManagedProfile = um.isManagedProfile(appInfo.userHandle.getIdentifier());
diff --git a/src/com/android/settings/nfc/PaymentBackend.java b/src/com/android/settings/nfc/PaymentBackend.java
index de733081a1d..021d673e152 100644
--- a/src/com/android/settings/nfc/PaymentBackend.java
+++ b/src/com/android/settings/nfc/PaymentBackend.java
@@ -36,7 +36,6 @@ import com.android.internal.content.PackageMonitor;
 
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Objects;
 
 public class PaymentBackend {
     public static final String TAG = "Settings.PaymentBackend";
@@ -53,24 +52,6 @@ public class PaymentBackend {
         public ComponentName settingsComponent;
         public UserHandle userHandle;
         public Drawable icon;
-
-        public String getKey() {
-            return Integer.toString(hashCode());
-        }
-
-        @Override
-        public int hashCode() {
-            return Objects.hash(componentName, userHandle);
-        }
-
-        @Override
-        public boolean equals(Object o) {
-            if (o == this) return true;
-            if (!(o instanceof PaymentAppInfo)) return false;
-            PaymentAppInfo appInfo = (PaymentAppInfo) o;
-            return componentName.equals(appInfo.componentName)
-                    && userHandle.equals(appInfo.userHandle);
-        }
     }
 
     /**
-- 
2.53.0

